website/src/_posts/2017-02-13-stateful-processing.md

```diff
@@ -499,31 +543,59 @@ new DoFn<KV<UserId, Event>, KV<UserId, Prediction>>() {
 ```
 
 ```py
-# State and timers are not yet supported in Beam's Python SDK.
-# Watch this space!
+class ModelStatefulFn(beam.DoFn):
+
+  PREVIOUS_PREDICTION = BagStateSpec('previous_pred_state', PredictionCoder())
+  MODEL_STATE = CombiningValueStateSpec('model_state',
+                                        ModelCoder(),
+                                        ModelFromEventsFn())
+
+  def process(self,
+              user_event,
+              previous_pred_state=beam.DoFn.StateParam(PREVIOUS_PREDICTION),
+              model_state=beam.DoFn.StateParam(MODEL_STATE)):
+    user = user_event[0]
+    event = user_event[1]
+    model = model_state.read()
+    previous_prediction = previous_pred_state.read()
+
+    new_prediction = model.prediction(event)
+    model_state.add(event)
+
+    if (previous_prediction is None
+        or self.should_output_prediction(
+            previous_prediction, new_prediction)):
+      previous_pred_state.clear()
+      previous_pred_state.add(new_prediction)
+      yield (user, new_prediction)
 ```
 
 Let's walk through it,
 
- - You have two state cells declared, `@StateId("model")` to hold the current
-   state of the model for a user and `@StateId("previousPrediction")` to hold
+ - You have two state cells declared,
+   `@StateId("model")` / `StateParam(MODEL_STATE)` to hold the current
+   state of the model for a user and
+   `@StateId("previousPrediction")/StateParam(PREVIOUS_PREDICTION)` to hold
    the prediction output previously.
  - Access to the two state cells by annotation in the `@ProcessElement` method
-   is as before.
- - You read the current model via `modelState.read()`. Because state is also
-   per-key-and-window, this is a model just for the UserId of the Event
-   currently being processed.
+   is as before, and in Python by using an argument placeholder.
+ - You read the current model via `modelState.read()` / `model_state.read()`.
```

Comment:
I would omit the python version of the code in the prose. I think the mapping is clear enough, especially with the disclaimer above.

Reply:
I think it does no harm. Is it okay if I leave it?

Reply:
It harms the readability. Having both variations is not worth stuttering of repeating tiny variations in the code. (As Max says, we can use a span with langauge classes, or I would be OK with just leaving the Java version in the prose as it was originally written.)